src/components/LeverageSlider/LeverageSlider.tsx

```diff
@@ -1,30 +1,62 @@
 import cx from "classnames";
-import { BASIS_POINTS_DIVISOR, MAX_ALLOWED_LEVERAGE } from "config/factors";
 import Slider, { SliderTooltip, Handle } from "rc-slider";
 import "rc-slider/assets/index.css";
 import "./LeverageSlider.scss";
+import { range } from "lodash";
+import { useCallback, useEffect, useMemo } from "react";
 
-const leverageMarks = {
-  2: "2x",
-  5: "5x",
-  10: "10x",
-  15: "15x",
-  20: "20x",
-  25: "25x",
-  30: "30x",
-  35: "35x",
-  40: "40x",
-  45: "45x",
-  50: "50x",
-};
+const defaultMarks = [2, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50];
+const DEFAULT_LEVERAGE_KEY = 20;
 
 type Props = {
   isPositive?: boolean;
   value?: number;
   onChange: (value: number) => void;
+  marks: number[];
+};
+
+type HandleProps = {
+  value: number;
+  dragging: boolean;
+  index: number;
+  keyValueMap: { [key: number]: number };
 };
 
+function getMarksWithLabel(marks: number[]) {
+  return marks.reduce((marks, value, index) => {
+    marks[index * 10] = `${value}x`;
+    return marks;
+  }, {} as { [key: number]: string });
+}
+
 export function LeverageSlider(p: Props) {
+  const { onChange, value, marks } = p;
+  const finalMarks = marks ?? defaultMarks;
+
+  const { marksLabel, keyValueMap, valueKeyMap } = useMemo(() => {
+    const marksLabel = getMarksWithLabel(finalMarks);
+    const { keyValueMap, valueKeyMap } = generateKeyValueMap(finalMarks);
+    return { marksLabel, keyValueMap, valueKeyMap };
+  }, [finalMarks]);
+
+  const defaultValue = valueKeyMap[value ?? 0] ?? DEFAULT_LEVERAGE_KEY;
+  const max = (finalMarks.length - 1) * 10;
+
+  const handleChange = useCallback(
+    (value: number) => {
+      onChange(keyValueMap[value ?? 0] ?? DEFAULT_LEVERAGE_KEY);
+    },
+    [onChange, keyValueMap]
+  );
+
+  useEffect(() => {
+    if (value !== defaultValue) {
+      handleChange(defaultValue);
+    }
+  }, [value, defaultValue, handleChange]);
+
+  const customHandle = (props: any) => <LeverageSliderHandle {...props} keyValueMap={keyValueMap} />;
```

Comment:
Just a minor rerender optimization idea: what do you think about wrapping the customHandle into a useMemo hook?

Reply:
Thanks, updated!